### PR TITLE
Updates for mParticle v7 / Branch v2-Event

### PR DESCRIPTION
**Pull request by E-B-Smith**
* Added support for the new mParticle v7 release.
* Added support for Branch v2-Event reporting.
* Added a sample app for testing: Example/mParticle-Branch-Example
* Added the mParticle switch that turns on Apple Search Ad checking.
* Added the optOut option to the integration.

The crucial integration file is here:
https://github.com/mparticle-integrations/mparticle-apple-integration-branchmetrics/pull/10/files#diff-19e479710f3b2f9cb2b27a0530efa50b

### CODE REVIEW

File: ChangeLog.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# mparticle-apple-integration-branchmetrics Change Log
+
+- v7.2.2
+  * Released April 20, 2018
```
Comment (samdozor) [inline ChangeLog.md]: Perhaps set this to v7.3.1, released May 2nd

File: Examples/mParticle-Branch-Example/mParticle/Info.plist [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		<string>UIInterfaceOrientationPortraitUpsideDown</string>
+	</array>
+	<key>UISupportedInterfaceOrientations~ipad</key>
+	<array>
+		<string>UIInterfaceOrientationPortrait</string>
+		<string>UIInterfaceOrientationPortraitUpsideDown</string>
+		<string>UIInterfaceOrientationLandscapeLeft</string>
+		<string>UIInterfaceOrientationLandscapeRight</string>
+	</array>
+	<key>branch_key</key>
```
Comment (samdozor) [inline Examples/mParticle-Branch-Example/mParticle/Info.plist]: the branch keys should not be in the plist.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 3]
```diff
@@ -18,186 +18,579 @@
...
-
 #if TARGET_OS_IOS == 1 && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
-    #import <UserNotifications/UserNotifications.h>
-    #import <UserNotifications/UNUserNotificationCenter.h>
+#import <UserNotifications/UserNotifications.h>
+#import <UserNotifications/UNUserNotificationCenter.h>
 #endif
 
+__attribute__((constructor))
+void MPKitBranchMetricsLoadClass(void) {
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m:27]: why was this needed? `load` should be called by the obj-c runtime without this?

File: mParticle-BranchMetrics/MPKitBranchMetrics.h [hunk 4]
```diff
@@ -23,10 +23,21 @@
 #import "mParticle.h"
 #endif
 
+extern void MPKitBranchMetricsLoadClass(void)
+    __attribute__((constructor));
+
 @interface MPKitBranchMetrics : NSObject <MPKitProtocol>
 
+// mParticle version 6 start:
+- (instancetype _Nonnull)initWithConfiguration:(NSDictionary*_Nonnull)configuration
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.h]: You shouldn't need the version 6 start in here - just the version 7. We maintain separate releases/branches for the two.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 5]
```diff
@@ -18,186 +18,579 @@
...
+    NSArray<NSString*>*_branchEventActions;
+    NSSet<NSString*>*_branchCategories;
 }
 
++ (nonnull NSNumber *)kitCode;
+
+- (instancetype _Nonnull) init NS_DESIGNATED_INITIALIZER;
+
+// Version 6 Start:
+- (instancetype _Nonnull)initWithConfiguration:(nonnull NSDictionary *)configuration
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m]: please remove.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 6]
```diff
@@ -18,186 +18,579 @@
...
+#pragma mark - MPKitInstanceProtocol Lifecycle Methods
 
+- (instancetype _Nonnull) init {
+    self = [super init];
+    self.configuration = @{};
+    self.launchOptions = @{};
+    return self;
+}
+
+- (instancetype _Nonnull)initWithConfiguration:(nonnull NSDictionary *)configuration
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m]: please remove

File: mParticle-BranchMetrics.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -264,10 +280,6 @@
 				DYLIB_COMPATIBILITY_VERSION = 1;
 				DYLIB_CURRENT_VERSION = 1;
 				DYLIB_INSTALL_NAME_BASE = "@rpath";
-				FRAMEWORK_SEARCH_PATHS = (
```
Comment (samdozor) [inline mParticle-BranchMetrics.xcodeproj/project.pbxproj]: `FRAMEWORK_SEARCH_PATHS` are necessary in order for Carthage to build.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 8]
```diff
@@ -18,186 +18,579 @@
...
+
+@interface MPEvent (Branch)
+- (MPMessageType) messageType;
+@end
+
 NSString *const ekBMAppKey = @"branchKey";
 NSString *const ekBMAForwardScreenViews = @"forwardScreenViews";
+NSString *const ekBMAEnableAppleSearchAds = @"enableAppleSearchAds";
+
+#pragma mark - Branch Helper - Move to Branch SDK
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m]: can you remove this if it's no longer necessary?

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 9]
```diff
@@ -18,186 +18,579 @@
...
 
 @interface MPKitBranchMetrics() {
-    Branch *branchInstance;
-    BOOL forwardScreenViews;
+    NSArray<NSString*>*_branchEventTypes;
+    NSArray<NSString*>*_branchEventActions;
+    NSSet<NSString*>*_branchCategories;
 }
 
++ (nonnull NSNumber *)kitCode;
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m:48]: Is there a reason all of these were added here? They're already defined in KitProtocol

File: mParticle-BranchMetrics/MPKitBranchMetrics.m [hunk 10]
```diff
@@ -18,186 +18,579 @@
...
-    MPKitExecStatus *execStatus = nil;
+- (MPKitExecStatus*) execStatus:(MPKitReturnCode)returnCode {
+    return [[MPKitExecStatus alloc] initWithSDKCode:self.class.kitCode returnCode:returnCode];
+}
+
+#pragma mark - MPKitInstanceProtocol Lifecycle Methods
 
+- (instancetype _Nonnull) init {
+    self = [super init];
+    self.configuration = @{};
```
Comment (samdozor) [inline mParticle-BranchMetrics/MPKitBranchMetrics.m]: setting these to an empty object *shouldn't* be a problem but we haven't tested it..is there a reason it was necessary?

**Review by samdozor — CHANGES_REQUESTED** [review]
First off - thank you so much for this extensive PR and the new sample app. I've left a number of comments inline :)

File: ChangeLog.md (hunk 1 above)
Comment (E-B-Smith) [reply ChangeLog.md]: Done.

File: Examples/mParticle-Branch-Example/mParticle/Info.plist (hunk 2 above)
Comment (E-B-Smith) [reply Examples/mParticle-Branch-Example/mParticle/Info.plist]: Done.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 3 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m:27]: In some project configurations, the runtime linker won't figure out that `MPKitBranchMetrics` is a dependency, so it won't load `MPKitBranchMetrics`, and the `+ load` method won't be called.

This is a sort of dumb work around to make sure the class is loaded. Since this function lives in the same file as the class implementation, the linker will load everything in the file.

File: mParticle-BranchMetrics/MPKitBranchMetrics.h (hunk 4 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.h]: I'll remove this.  But I haven't done any work on the v6 version. I was hoping that the same integration would work for v6 and v7.  What's the life expectancy of v6?

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 5 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m]: Done.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 6 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m]: Done.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 8 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m]: Done.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 10 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m]: `self.configuration` is marked `nonnull` in the header, so I assigned it a non-null value.

File: mParticle-BranchMetrics.xcodeproj/project.pbxproj (hunk 7 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics.xcodeproj/project.pbxproj]: Re-added.

File: mParticle-BranchMetrics/MPKitBranchMetrics.h (hunk 4 above)
Comment (samdozor) [reply mParticle-BranchMetrics/MPKitBranchMetrics.h]: v6 is deprecated and all customers are upgrading to v7 - migration started last year and should be complete shortly. I propose only making this change for v7.

File: mParticle-BranchMetrics/MPKitBranchMetrics.m (hunk 9 above)
Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m:48]: It's listed in the `MPKitProtocol` protocol as a required method, and I think it was in the v6 version of the code.  I can remove it.  Your call.

Comment (samdozor) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m:48]: Sorry I just mean why re-define all of these interfaces?

```
+ (nonnull NSNumber *)kitCode;

- (instancetype _Nonnull) init NS_DESIGNATED_INITIALIZER;

// Version 7 Start:
- (MPKitExecStatus*_Nonnull)didFinishLaunchingWithConfiguration:(nonnull NSDictionary *)configuration;

- (void)start;

- (MPKitExecStatus*_Nonnull)continueUserActivity:(nonnull NSUserActivity *)userActivity
    restorationHandler:(void(^ _Nonnull)(NSArray * _Nullable restorableObjects))restorationHandler;

- (MPKitExecStatus*_Nonnull)openURL:(nonnull NSURL *)url
                            options:(nullable NSDictionary<NSString *, id> *)options;

- (MPKitExecStatus*_Nonnull)openURL:(nonnull NSURL *)url
                  sourceApplication:(nullable NSString *)sourceApplication
                         annotation:(nullable id)annotation;

- (MPKitExecStatus*_Nonnull)receivedUserNotification:(nonnull NSDictionary *)userInfo;
- (MPKitExecStatus*_Nonnull)logCommerceEvent:(nonnull MPCommerceEvent *)commerceEvent;
- (MPKitExecStatus*_Nonnull)logEvent:(nonnull MPEvent *)event;
- (MPKitExecStatus*_Nonnull)setKitAttribute:(nonnull NSString *)key value:(nullable id)value;
- (MPKitExecStatus*_Nonnull)setOptOut:(BOOL)optOut;
```

and not just have the implementations?

Comment (E-B-Smith) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m:48]: So I would get a compiler warning if I forgot to implement one of them.

Comment (samdozor) [reply mParticle-BranchMetrics/MPKitBranchMetrics.m:48]: ah ok, no worries then.

**Comment by E-B-Smith** [comment]
Thanks!
